Wrote an intro based on brainstorm notes (#11)
chhoumann/p7: report/sections/Chapter 1/Intro.tex

\chapter{Introduction} \label{chap:introduction}
On the 1st semester of the Software Master’s Degree at \aau{} the programming paradigms course currently has a problem where most of the attending students failed the course last time it was held. This led to a large survey being sent out and a student task force being assembled to grasp the main flaws with how the course was structured. Here a large part of the students mentioned a lack of expertise in the programming languages taught in the course and a lack of possibilities for actively training their skills during the semester. 

> With this knowledge in mind, \aau{} chose to restructure the course with a larger focus on practical programming as opposed to the more classic lectures that are usually used at universities. To support this way of teaching, the course lectures have been moved to a more classical classroom setup but with a monitor at each table and a main monitor at the lecture desk. This is meant to allow the students to more easily work together in groups while working on the shared monitor and allowing the lectures to view and share the groups monitors allowing actual implementations by the students to be used for classroom discussions. 

Reviewer: ~~knowledge~~
"Restructure the course entirely. The course, which previously introduced the students to a large number of languages" 
What is the actual problem? Right now it does not seem that we try to fix a problem but rather accomodate changes.

> This way of solving the problem brings its problems with it, however. As all students are using their own local environments for solving the exercises at hand, other students cannot revisit the exercise solutions discussed as these are not shared, making it hard to revisit them later. Furthermore, it can be hard to remember concrete implementations as well as verify whether your own implementation is correct and lives up to the requirements of the exercise. In fact, it is not uncommon for people to accidentally misunderstand an exercise and therefore implement the wrong solution.  

Reviewer: "its problem" It is unclear what 'its' refer to.  "problem brings its problems".

What are environments?

\section{Initial Problem} 
\aau's current restructuring of the course already seems like a major improvement over the former structure but as mentioned it still suffers from a few notable problems.
\begin{itemize}
	\item No way to revisit discussed solutions
	\item No way to verify that an implementation is correct
	\item No way for the entire group to work together on the same code
	\item No way for the professor to verify how far the students are in the problem solving
	\begin{itemize}
		\item Hard to verify that all students have solved the problems
		\item Hard to judge the skill-level of the students and if the exercises are suitable
	\end{itemize}
\end{itemize}

> To try and solve these problems we would like to develop a website targeted toward the professor that enables him to gain a better overview during exercise sessions and therefore allows him to improve the quality of the sessions. To facilitate this, we imagine a website that should allow the professor to:

Reviewer: This is good. However, we need to fix one problem not multiple.

> \begin{itemize}

Reviewer: It discusses a solution before the problem is properly defined.
- We need the problem to be properly established, and the properties of the solution described before the solution itself. 
- The problem is xxxx
- The solution needs to be scalable to accomodate the many students and xxxxx

	\item The professor should be able to create a syllabus with the exercises that students are expected to solve
	\begin{itemize}
		\item The professor should be able to create multiple syllabi where students can self-enrol into the relevant ones for them
	\end{itemize}
	\item The professor should be able to track the amount of exercises completed by each enrolled student
	\begin{itemize}
		\item The professor should be able to view additional stats such as attempts used 
	\end{itemize}
	\item The professor should be able to specify criteria in order to pass each exercise 
	\begin{itemize}
		\item The students should be able to verify is their solution lives up to the specified criteria
		\item Criteria checking should be handled automatically by the system
	\end{itemize}
	\item The students should be able to collaborate on the same code
	\item The students should have access to the discussed solutions later on
	\item The professor should be able to see each students code
\end{itemize}

> \subsection{MVP}

Reviewer: This needs to be after a proper problem statement.

The presented features were then prioritized by a group of students to construct a minimal viable product (MVP) that could be tested and used to verify if the presented product would provide anything useful to the course. Since the MVP is meant to be used as a demo product that allows for additional input to be gathered only the minimum number of functionalities needed were included.

> \begin{displayquote}

Reviewer: This comes too late. Does it need to be a website? That is an implementation detail to the solution? It is unlcear why the erify part is needed to solve the problem.

\chapter{Introduction} \label{chap:introduction}

> On the 1st semester of the Software Master’s Degree at \aau{} the Programming Paradigms course currently has a problem where most of the attending students failed the course last time it was held. This led to a survey being sent out and a student task force being assembled to grasp the main flaws with how the course was structured. Here a large part of the students mentioned a lack of expertise in the programming languages taught in the course and a lack of possibilities for actively training their skills during the semester. 

Reviewer 2: ```suggestion
On the 1st semester of the Software Master’s Degree at \aau{}, the Programming Paradigms course had a problem where most attending students failed the course last year. 
This resulted in a survey being sent out, and a student task force being assembled to grasp the main flaws with how the course was structured. 
Here, many students mentioned lacking experience in the programming languages taught in the course given few opportunities to actively train their skills during the semester. 
```

With this in mind, \aau{} chose to restructure the course entirely. The course now focuses on a single programming language instead of multiple. Furthermore, a larger focus is on practical programming as opposed to the more classic lectures that are usually used at universities. To support this way of teaching, the course lectures have been moved to a more classical classroom setup but with a monitor at each table and a main monitor at the lecture desk. This is meant to allow the students to more easily work together in groups while working on the shared monitor and allowing the lectures to view and share the groups monitors allowing actual implementations by the students to be used for classroom discussions. 

> This way of solving the pass rate problem brings its problems with it, however. As all students are using their own local development environments for solving the exercises at hand, other students cannot revisit the exercise solutions discussed as these are not shared, making it hard to revisit them later. Furthermore, it can be hard to remember concrete implementations as well as verify whether your own implementation is correct and lives up to the requirements of the exercise. In fact, it is not uncommon for people to accidentally misunderstand an exercise and therefore implement the wrong solution.  

Reviewer 2: ```suggestion
This new approach causes its own problems, however. 
Since students use their own local development environments for solving exercises, other students cannot revisit the discussed exercise solutions as these are not shared, making it impossible to revisit them later. 
Furthermore, it is difficult to verify whether your own solution is correct and lives up to the requirements of the exercise. 
In fact, it is not uncommon for people to accidentally misunderstand an exercise and therefore implement an incorrect solution.
```

> \section{Initial Problem} \label{sec:Initial Problem}

Reviewer 2: ```suggestion
\section{Initial Problem} \label{sec:initial-problem}
```

> \aau's current restructuring of the course already seems like a major improvement over the former structure but as mentioned it still suffers from a few notable problems all related to the execution exercise sessions:

Reviewer 2: ```suggestion
The restructuring of the course already seems like a major improvement over the former structure, but as mentioned it still suffers from a few notable problems related to the execution of the exercise sessions:
```

\begin{itemize}
	\item No way to revisit discussed solutions

> 	\item No way to verify that an implementation is correct

Reviewer 2: ```suggestion
	\item No way to verify that your own solution is correct
```

	\item No way for the entire group to work together on the same code

> 	\item No way for the professor to verify how far the students are in the problem solving

Reviewer 2: ```suggestion
	\item No way for the professor to verify how many exercises the students have solved
```

	\begin{itemize}
		\item Hard to verify that all students have solved the problems

> 		\item Hard to judge the skill-level of the students and if the exercises are suitable

Reviewer 2: ```suggestion
		\item Difficult to judge the skill level of the students and whether the difficulty of exercises is suitable
```

	\end{itemize}
\end{itemize}

> Summarising the main problems that have risen after the switch to a more practical exercise-focused teaching environment an initial problem statement can be written.

Reviewer 2: ```suggestion
By summarizing the main problems following the switch to a more practical and exercise focused teaching environment, an initial problem statement can be written.
```

\begin{displayquote}

> How can a tool which facilitates a more streamlined user experience for the students while functioning as a feedback loop for the professor allowing them to better structure the course based on statistics gathered from the students' exercise statistics be developed?

Reviewer 2: ```suggestion
How can a tool be developed which facilitates a more streamlined user experience for the students while functioning as a feedback loop for the professor allowing them to better structure the course based on statistics gathered from the students' exercise statistics?
```

\end{displayquote} 

\subsection{Minimal Viable Product}

> Considering that the students attending the course are all using different Operating systems the simplest way to develop a tool that would be available for all would be to create a web application. To facilitate the problems mentioned in \ref{sec:Initial Problem} we imagine a website that allows the users to complete the following tasks:

Reviewer 2: ```suggestion
Considering that students use a variety of different operating systems, a web application would be the best way to ensure easy portability.
To facilitate the problems mentioned in \ref{sec:initial-problem}, we imagine a website that allows users to complete the following tasks:
```

\begin{itemize}

> 	\item The professor should be able to create a syllabus with the exercises that students are expected to solve

Reviewer 2: ```suggestion
	\item The professor should be able to create a syllabus with exercises that students are expected to solve
```

	\begin{itemize}
		\item The professor should be able to create multiple syllabi where students can self-enrol into the relevant ones for them
	\end{itemize}

> 	\item The professor should be able to track the amount of exercises completed by each enrolled student

Reviewer 2: ```suggestion
	\item The professor should be able to track the amount of exercises solved by each enrolled student
```

	\begin{itemize}

> 		\item The professor should be able to view additional stats such as attempts used 

Reviewer 2: ```suggestion
		\item The professor should be able to view additional statistics like the number of attempts used by a student to solve an exercise
```

	\end{itemize}
	\item The professor should be able to specify criteria in order to pass each exercise 
	\begin{itemize}

> 		\item The students should be able to verify is their solution lives up to the specified criteria

Reviewer 2: ```suggestion
		\item The students should be able to verify whether their solution is correct
```

		\item Criteria checking should be handled automatically by the system
	\end{itemize}

> 	\item The students should be able to collaborate on the same code

Reviewer 2: ```suggestion
	\item Students should be able to collaborate on the same code
```

> 	\item The students should have access to the discussed solutions later on

Reviewer 2: ```suggestion
	\item Students should have access to the discussed solutions later on
```

> 	\item The professor should be able to see each students code

Reviewer 2: ```suggestion
	\item The professor should be able to see each student's code
```

\end{itemize}

> The presented features were then prioritized by a group of students to construct a minimal viable product (MVP) that could be tested and used to verify if the presented product would provide anything useful to the course. Since the MVP is meant to be used as a demo product that allows for additional input to be gathered only the minimum number of functionalities needed were included.

Reviewer 2: ```suggestion
We then prioritized the presented features required to construct a minimal viable product (MVP).
We plan on using this in practice to see whether it serves as a useful solution to the presented problems. 
The MVP is meant to be used as a demo product for gathering additional feedback.
As such, only a minimal number of features are included in our proposed solution:
```

\begin{displayquote}

> A website that allows a user to specify a set of exercises as well as the completion criteria for each of these exercises. Furthermore an option to verify whether the submitted code is a correct solution to the chosen exercise.

Reviewer 2: ```suggestion
A website that allows a professor to specify a set of exercises as well as the completion criteria for each of these exercises. Furthermore, a way to verify whether submitted code is a correct solution to the given exercise.
```

\chapter{Introduction} \label{chap:introduction}
On the 1st semester of the Software Master’s Degree at \aau{}, the Programming Paradigms course had a problem where most attending students failed the course last year. 
This resulted in a survey being sent out, and a student task force being assembled to grasp the main flaws with how the course was structured. 
Here, many students mentioned lacking experience in the programming languages taught in the course given few opportunities to actively train their skills during the semester. 

With this in mind, \aau{} chose to restructure the course entirely. 
The course now focuses on a single programming language instead of multiple. 
Furthermore, the course features more practical programming instead of following the traditional lecture model usually used at universities. 
To support this way of teaching, the course lectures have been moved to a more classical classroom setup, but with a monitor at each table and a main monitor at the lecture desk. 
This is meant to allow students to more easily work together in groups by using the shared monitor.

> In addition, it allows the professor to view and share other groups' monitors such that student solutions can used for classroom discussions. 

Reviewer 3: ```suggestion
In addition, it allows the professor to view and share other groups' monitors such that student solutions can be used for classroom discussions. 
```

This resulted in a survey being sent out, and a student task force being assembled to grasp the main flaws with how the course was structured. 
Here, many students mentioned lacking experience in the programming languages taught in the course given few opportunities to actively train their skills during the semester. 

With this in mind, \aau{} chose to restructure the course entirely. 
The course now focuses on a single programming language instead of multiple. 
Furthermore, the course now features more practical programming exercises with classroom discussion in-between, rather than traditional lectures followed by a seperate exercise session.
To support this way of teaching, the course lectures have been moved to a more classical classroom setup, but with a monitor at each table and a main monitor at the lecture desk. 
This is meant to allow students to more easily work together in groups by using the shared monitor.
In addition, it allows the professor to view and share other groups' monitors such that student solutions can be used for classroom discussions. 

> This new approach causes its own problems, however. 

Reviewer 3: ```suggestion
However, with this way of teaching new problems arise.
```

> Since students use their own local development environments for solving exercises, other students cannot revisit the discussed exercise solutions as these are not shared, making it impossible to revisit them later. 

Reviewer 3: ```suggestion
Since students use their own local development environments for programming exercises, other students cannot revisit the discussed exercise solutions created by other groups. This makes it difficult to reflect on and learn from other students' solutions.
```

Furthermore, it is difficult to verify whether your own solution is correct and lives up to the requirements of the exercise. 

> In fact, it is not uncommon for people to accidentally misunderstand an exercise and therefore implement an incorrect solution.

Reviewer 4: ```suggestion
In fact, it is not uncommon for people to misunderstand an exercise and therefore implement an incorrect solution.
```

Misunderstandings are always accidental.

\section{Initial Problem} \label{sec:initial-problem}
The restructuring of the course already seems like a major improvement over the former structure, but as mentioned it still suffers from a few notable problems related to the execution of the exercise sessions:
\begin{itemize}
	\item No way to revisit discussed solutions
	\item No way to verify that your own solution is correct
	\item No way for the entire group to work together on the same code
	\item No way for the professor to verify how many exercises the students have solved
	\begin{itemize}
		\item Difficult to verify that all students have solved the problems
		\item Difficult to judge the skill level of the students and whether the difficulty of exercises is suitable
	\end{itemize}
\end{itemize}

> By summarizing the main problems following the switch to a more practical and exercise focused teaching environment, an initial problem statement can be written.

Reviewer 4: ```suggestion
By summarizing the main problems following the switch to the new teaching approach, an initial problem statement is written.
```

\begin{displayquote}

> How can one ensure a mutual feedback loop between students, TAs and the professor during exercise session?

Reviewer 4: Det eneste vi rigtigt har beskrevet som et problem (som ikke bare er en liste der påstår problemer) er misforståelse af opgaver og difficulty in verifying solution og finde game solutions frem. Mutual feedback kommer som en overraskelse for mig som læser.

\end{displayquote} 

\subsection{Minimal Viable Product}
Considering that students use a variety of different operating systems, a web application would be the best way to ensure easy portability.
To facilitate the problems mentioned in \ref{sec:initial-problem}, we imagine a website that allows users to complete the following tasks:

> \begin{itemize}

Reviewer 4: Det er måske en smule tidligt at snakke om hvad hjemmesiden skal kunne.

Author: The subsection can be moved, later on, I do however feel like it gives a decent view of what we're going to do now for Carlos
